File: solutions/javascript/mixed-juices/1/mixed-juices.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Determines how long it takes to prepare a certain juice.
 *
 * @param {string} name
 * @returns {number} time in minutes
 */
export const timeToMixJuice = (name) => {
    switch (name) {
        case "Pure Strawberry Joy":
            return 0.5;
        case "Energizer":
        case "Green Garden":
            return 1.5;
        case "Tropical Island":
            return 3;
        case "All or Nothing":
            return 5;
        default:
            return 2.5;
    }
};

/**
 * Calculates the number of limes that need to be cut
 * to reach a certain supply.
 *
 * @param {number} wedgesNeeded
 * @param {string[]} limes
 * @returns {number} number of limes cut
 */
export const limesToCut = (wedgesNeeded, limes) => {
    let wedgesCut = 0;
    let limesCut = 0;
    let limesLength = limes.length;

    while (wedgesCut < wedgesNeeded && limesCut < limesLength) {
        const lime = limes[limesCut];
        wedgesCut += getWedgeValue(lime);
        limesCut += 1;
    }

    return limesCut;

    function getWedgeValue(/** @type {string} lime*/ lime) {
        switch (lime) {
            case "small":
                return 6;
            case "medium":
                return 8;
            case "large":
                return 10;
            default:
                return 0;
        }
    }
};

/**
 * Determines which juices still need to be prepared after the end of the shift.
 *
 * @param {number} timeLeft
 * @param {string[]} orders
 * @returns {string[]} remaining orders after the time is up
 */
export const remainingOrders = (timeLeft, orders) => {};

File: solutions/julia/mixed-juices/1/mixed-juices.jl
function time_to_mix_juice(juice::String)::Float64
    mix_times = Dict{String,Float64}(
        "Pure Strawberry Joy" => 0.5,
        "Energizer" => 1.5,
        "Green Garden" => 1.5,
        "Tropical Island" => 3.0,
        "All or Nothing" => 5.0
    )

    get(mix_times, juice, 2.5)
end

function wedges_from_lime(size)
    wedge_counts = Dict("small" => 6, "medium" => 8, "large" => 10)
    get(wedge_counts, size, 0)  # Returns 0 for unknown sizes
end

function limes_to_cut(needed, limes)
    limes_cut = 0
    limes = copy(limes)

    while needed > 0 && !isempty(limes)
        limes_cut += 1
        needed -= wedges_from_lime(popfirst!(limes))
    end

    limes_cut
end



function order_times(orders)
    time_to_mix_juice.(orders)
end

function remaining_orders(time_left, orders)
    orders = copy(orders)

    while time_left > 0 && !isempty(orders)
        time_left -= time_to_mix_juice(first(orders))
        popfirst!(orders)
    end

    orders
end
